pkg/retriever/coordinators/race.go

```diff
@@ -2,56 +2,51 @@ package coordinators
 
 import (
 	"context"
-	"errors"
+	"fmt"
 	"sync"
 
 	"github.com/filecoin-project/lassie/pkg/types"
 	"go.uber.org/multierr"
 )
 
-// Race retrieves by racing one or more canditate retrievals together, taking the first successful result or returning a combined error if all fail
-func Race(ctx context.Context, retrievalCalls []types.CandidateRetrievalCall) (*types.RetrievalStats, error) {
+func Race(ctx context.Context, queueOperations types.QueueRetrievalsFn) (*types.RetrievalStats, error) {
 	resultChan := make(chan types.RetrievalResult)
 	ctx, cancel := context.WithCancel(ctx)
-
+	defer cancel()
 	var waitGroup sync.WaitGroup
-	waitGroup.Add(len(retrievalCalls))
-	for _, retrievalCall := range retrievalCalls {
-		retrievalCall := retrievalCall
-		go func() {
-			defer waitGroup.Done()
-			stats, err := retrievalCall.CandidateRetrieval.RetrieveFromCandidates(retrievalCall.Candidates)
-			select {
-			case resultChan <- types.RetrievalResult{Stats: stats, Err: err}:
-			case <-ctx.Done():
-			}
-		}()
-	}
-	stats, err := collectResults(ctx, resultChan, len(retrievalCalls))
-	cancel()
-	return stats, err
-}
-
-func collectResults(ctx context.Context, resultChan <-chan types.RetrievalResult, totalRetrievers int) (*types.RetrievalStats, error) {
+	waitGroup.Add(1)
+	go func() {
+		defer waitGroup.Done()
+		queueOperations(ctx, func(nextRetrieval types.RetrievalTask) {
+			waitGroup.Add(1)
+			go func() {
+				defer waitGroup.Done()
+				stats, err := nextRetrieval.Run()
+				select {
+				case <-ctx.Done():
```

Comment:
`ctx` might get cancelled, but unless the callback inejects it into `nextRetrieval` then we're in trouble aren't we?

perhaps `Run()` should take a `ctx`, then the coordinator could at least get its context closer to the action rather than relying on the external party to wire that up.

or .. what am I missing?

Reply:
the retrieval itself should use the Ctx passed to the queueOperations to setup it's task. other than the tests, everything should be doing this.

Reply:
I agree maybe it makes better sense in the long term for ctx to go directly to run.